### ISSUE TITLE
bug(callFirestore): calling callFirestore with an object does not work on Windows (only fixtures do)

### ISSUE THREAD

**Issue by dhair-seva** [prescottprue/cypress-firebase#71]
I tried calling:
```
cy.callFirestore("update", "companies/my-id", { creditsRemaining: 1000 })
```
and
```
cy.callFirestore("set", "companies/my-id", { creditsRemaining: 1000 })
```
But receive this error:
```
Error: the string "(node:25288) UnhandledPromiseRejectionWarning: Error: Fixture not found at path: test\\e2e\\fixtures\\'{creditsRemaining:1000}' or cypress\\fixtures\\'{creditsRemaining:1000}'\n    at readFixture (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\firebase-tools-extra\\lib\\commands\\firestore.js:26:19)\n    at C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\firebase-tools-extra\\lib\\commands\\firestore.js:96:39\n    at step (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\tslib\\tslib.js:136:27)\n    at Object.next (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\tslib\\tslib.js:117:57)\n    at C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\tslib\\tslib.js:110:75\n    at new Promise (<anonymous>)\n    at Object.__awaiter (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\tslib\\tslib.js:106:16)\n    at firestoreAction (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\firebase-tools-extra\\lib\\commands\\firestore.js:87:20)\n    at runFirebaseExtra (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\firebase-tools-extra\\bin\\firebase-tools-extra:57:12)\n    at Object.<anonymous> (C:\\Users\\I20649\\Workspace\\testramp-ng8\\node_modules\\firebase-tools-extra\\bin\\firebase-tools-extra:66:2)\n(node:25288) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)\n(node:25288) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code." was thrown, throw an Error :)

Because this error occurred during a 'before all' hook we are skipping the remaining tests in the current suite: 'Client New Test Run Page'
```
It looks like it won't allow an object like your documentation states:
```
cy.callFirestore('set', `testCollection/${TEST_UID}`, {
  name: 'axa',
  age: 8,
})
```

Versions:
**cypress-firebase:** "0.4.1"
**cypress:** "3.4.1"
**firebase-tools:** "^7.8.0"

This seems like a bug.

**Comment by prescottprue** [comment]
Thanks for reporting. Yeah it should support both

**Comment by dhair-seva** [comment]
Interesting, When I run this command manually, I get that same error:
`.\node_modules\.bin\npx firebase-extra firestore update 'companies/mycomp' '{"creditsRemaining":1000}' -P my-project`

So it seems that firebase-extra is the one with the issue?

**Comment by dhair-seva** [comment]
Actually, looks like the error is displayed here.
https://github.com/prescottprue/firebase-tools-extra/blob/master/src/commands/firestore.ts

**Comment by prescottprue** [comment]
@dhair-seva Did you try updating to v0.5.0? I tried to replicate and am not seeing that issue anymore with that version (see below), I didn't realize earlier when you posted that you were using 0.4.1.

When calling the following:
`cy.callFirestore('update', 'projects/123ABC', { some: 'data' })`

Result in logs is successful:
![image](https://user-images.githubusercontent.com/2992224/70210730-6ed3c580-1701-11ea-94db-e9553b52942d.png)

The same also works with set

**Comment by dhair-seva** [comment]
Interesting. I didn't realize there was a new version. I will try that now.

**Comment by dhair-seva** [comment]
I just updated and tried it in 0.5.0 and still receiving the same error.

**Comment by dhair-seva** [comment]
Could it be the firebase-extra command that we use is out of date on mine? I imagine not, but that seems like it would be the only reason...

**Comment by prescottprue** [comment]
What version of firebase-tools-extra is installed? Also, what about firebase-tools?

**Comment by dhair-seva** [comment]
**firebase-tools**: 7.8.0
**firebase-tools-extra**: 0.2.1

**Comment by dhair-seva** [comment]
Any ideas? This is really blocking a lot of my e2e testing. I would love to help contribute, but not sure where else to look for, on where the bug would be.

**Comment by dhair-seva** [comment]
Can we reopen this?

**Comment by prescottprue** [comment]
Not sure of how to replicate - On both Mac and Linux I am able to pass the object and it is handled correctly. Do you have anything setup, like a CI provider, that is experiencing this on linux?

I'll try replicating again, but as pictured above it seemed to be working last time I tried

**Comment by dhair-seva** [comment]
Perhaps it's the fact that I have Windows? I don't use Linux or Mac. There
may be a chance that I can send it up to circleci. However I haven't setup
that logic yet. Do you think it could be Windows. I'm quite certain it's
firebase-extra though because I get the same error on the command line.

**Comment by dhair-seva** [comment]
I will generate a new angular application with the same setup and see if I
can make a repro

**Comment by dhair-seva** [comment]
> Something to note though is that it doesn't seem that the `-D` flag was added on the command that was generated for you, where it was on mine when passed the same arguments, so I'm going to look into the logic for that

In looking at your screenshot of your command, I don't see a -D command. What does -D do?

**Comment by dhair-seva** [comment]
I'm getting this same command in my current PR `feature/dmh/fix-firebase-path`. The one that I have for another issue I logged. I ran this command:
`node_modules\.bin\firebase-extra firestore set "companies/DevAutomation/1575951379044" '{"_id":"1"}' -P testramp-dan`
I received the same error, so its not specific to my cypress app. I'm more convinced that its windows. I have a MAC computer though. I'm going to try the command

**Comment by prescottprue** [comment]
Yeah that sound right that it would be due to the OS - I haven't gotten the chance to test any of these tools on Windows

**Comment by dhair-seva** [comment]
Well, my team and I are on Windows and would love to get this thing more
stable on Windows. I have a temporary workaround as I stated but it's
slowing our tests down a lot. However I am really grateful for your
efforts. This module is a life saver.

**Comment by dhair-seva** [comment]
Okay, so I tried this same command on my MAC, and it works just fine. This is a Windows-specific bug.

**Comment by dhair-seva** [comment]
Any update on this? You have a tag for "Needs Repro", but this happens in any windows machine. Can we remove it?

**Comment by prescottprue** [comment]
I don't currently have a windows machine to test on, and it will be a little while until I can get a chance to spin up a vm for testing this. I'll remove the needs-repro label, but I will need to reproduce on a windows machine so I can make changes to fix the issue. Unless you would be open to making a PR?

**Comment by dhair-seva** [comment]
Sure, Id be happy too when I get some time. I assume this would be in your
other repo since this is when running the firebase extra command?

**Comment by prescottprue** [comment]
Great! Yes correct, it will most likely be in firebase-tools-extra

**Comment by Alberick** [comment]
Hello, I just started looking at this package. I'm also using windows and I'm facing the same scenario. This is what I get when it runs on cypress.

```
Console was cleared
cypress_runner.js:159056 Command:     exec
cypress_runner.js:159056 Yielded:     Objectcode: 1stderr: ""stdout: ""__proto__: Object
cypress_runner.js:159056 Error:       CypressError: cy.exec('npx firebase-extra firestore set config/tester '{"name":"tester"}' -P myProject') failed because the command exited with a non-zero code.

Pass {failOnNonZeroExit: false} to ignore exit code failures.

Information about the failure:
Code: 1


Because this error occurred during a 'before all' hook we are skipping the remaining tests in the current suite: 'Select/Deselect all queries'
cypress_runner.js:159056 Snapshot:    The snapshot is missing. Displaying current state of the DOM.
```

This works fine If I run on the cli ```npx firebase-extra firestore set config/tester '{"name":"tester"}'```

Do you have any advice or workaround I can use?

**Comment by dhair-seva** [comment]
I have a temporary workaround. Basically we save the object and then use that as a fixture. It works for our use case and we have used over 50 e2e tests with it. Obviously its not ideal, but it works for now.

```
import { isObject } from "lodash"
import * as uuid from "uuid/v4"

Cypress.Commands.overwrite("callFirestore", (originalFn: any, type: string, path: any, fixtureOrData: string | any, ...args) => {
  let id = ""
  if (isObject(fixtureOrData)) {
    id = (fixtureOrData as any)._id
    if (!id) {
      id = "Automation-" + uuid()
    }

    const fixturePath = "test-fixtures/" + id + ".json"
    return cy
      .writeFile("cypress/fixtures/" + fixturePath, fixtureOrData, "utf-8")
      .then(() => {
        fixtureOrData = fixturePath
        return originalFn(type, path, fixtureOrData, ...args)
      })
  }

  return originalFn(type, path, fixtureOrData, ...args)
})
```

**Comment by prescottprue** [comment]
I have started work on a new method of doing things (through tasks instead of passing through the firebase-tools-extra cli). This should help fix things, but it is definetly a work in progress

To try it out do the following:

1. Install from alpha and install firebase-admin `npm i --save-dev cypress-firebase@alpha firebase-admin` (currently at 0.11.0-alpha.2)
1. Switch your plugin import to be `pluginWithTasks` like so:
```diff
- const cypressFirebasePlugin = require('cypress-firebase').plugin
+ const cypressFirebasePlugin = require('cypress-firebase').pluginWithTasks
```
1. Pass the new arguments to the new plugin:
```diff
+ const admin = require('firebase-admin')
const cypressFirebasePlugin = require('cypress-firebase').pluginWithTasks

module.exports = (on, config) => {
+  return cypressFirebasePlugin(on, config, admin)
-  return cypressFirebasePlugin(config)
}
```

Full example

```js
const admin = require('firebase-admin')
const cypressFirebasePlugin = require('cypress-firebase').pluginWithTasks

module.exports = (on, config) => {
// Pass on function, config, and admin instance. Returns extended config
  return cypressFirebasePlugin(on, config, admin)
}
```

**Comment by prescottprue** [comment]
I have already noticed a few small things that should be fixed and improved - all changes are going out to the `alpha` tag for now, so feel free to just point to that in your package file if you want to keep getting updates of these versions while they progress

**Comment by prescottprue** [comment]
The v1.0.0 pre-release series now uses tasks by default and is being released on the `next` tag. Going to leave this open until it is released as latest.

Please reach out if you have any issues using the new plugin with tasks (available in 0.11.0) or using the new 1.0.0 pre-releases

**Comment by prescottprue** [comment]
v1.0.0 has been released to latest - everything is now going through tasks, so parsing should hopefully not be an issue regardless of os. Let me know if you get a chance to try it out!
